### DISCUSSION TITLE
How to parse from a byte* and length?

### DISCUSSION THREAD

**Discussion by fantassin** [jamescourtney/FlatSharp#405] [Q&A]
Hello,

I have large flatbuffer files and lots of them, ~20000 with a total size of ~100GB. I memory map each file and use Google's official flatbuffers .NET implementation to consume them.

I can't figure out how to create a `ReadOnlyMemory<byte>` or any other way to consume my memory mapped flatbuffers with `FlatSharp`.

Thank you.

**Comment by dvankley** [comment]
I have the same question. I'm using Steamworks.NET, which hands me pointers to message buffers in native memory. I want to deserialize them in Greedy mode because of my storage and read patterns, as well as the fact that I need to hand back the native memory buffer quickly. I don't think there's any real way of avoiding one copy and allocation (maybe pooling to avoid the allocation), but I can't figure out how to avoid an additional copy since FlatSharp won't accept a `Span` to deserialize from.

Deserializing from a `Span` doesn't make much sense for Lazy and Progressive modes, but it does seem like it should be supported for Greedy.

I could (and probably will, absent any other better ideas) implement my own `MemoryManager<byte>` to get a `Memory` view to the native memory, but that seems a bit janky.

**Comment by jamescourtney** [comment]
The `MemoryManager<byte>` approach is the best answer I'm aware of. It _is_ a little janky I agree, but the class is pretty simple to implement yourself. You could presumably pool the instances so you can save some allocations if that's a concern.

**Reply by jamescourtney** [reply]
One other thought here -- you can also consider implementing `IInputBuffer` yourself and passing in the pointer and length. One note here, however: FlatSharp does not perform bounds checking itself. Instead, it defers this to the various `IInputBuffer` implementations. When using `Memory`/`byte[]`/etc, this means the bounds checks are provided by .NET. This approach may result in a faster implementation since `Memory<byte>` does have some overhead.
